### PR TITLE
Fix propagation of an earlier exception in Transformer

### PR DESCRIPTION
**Pull request by makortel**
#### PR description:

This PR proposes a possible fix for https://github.com/cms-sw/cmssw/issues/40868.

#### PR validation:

`FWCore/Integration` unit tests pass

### CODE REVIEW

**Comment by makortel** [comment]
@cmsbuild, please test

**Comment by makortel** [comment]
@Dr15Jones What do you think?

File: FWCore/Integration/test/transformTest_cfg.py [hunk 1]
```diff
@@ -27,6 +28,8 @@
 process.maxEvents.input = 4
 
 process.start = cms.EDProducer("IntProducer", ivalue = cms.int32(1))
+if args.exception:
+  process.start = cms.EDProducer("FailingProducer")
```
Comment (Dr15Jones) [inline FWCore/Integration/test/transformTest_cfg.py:32]: So the test failure is in what is consumed and not in the actual transform call its.

File: FWCore/Framework/src/ProductResolvers.cc [hunk 2]
```diff
@@ -516,6 +516,17 @@ namespace edm {
     return Resolution(nullptr);
   }
 
+  void TransformingProductResolver::putProduct(std::unique_ptr<WrapperBase> edp) const {
+    // Override putProduct() to not set the resolver status to
+    // ResolveFailed when the Event::commit_() checks which produced
+    // products were actually produced and which not, because the
+    // transforming products are never produced by time of commit_()
+    // by construction.
+    if (edp) {
```
Comment (Dr15Jones) [inline FWCore/Framework/src/ProductResolvers.cc:525]: Should the putProduct call always get a null wrapper? Or does the later call also call putProduct?

File: FWCore/Integration/test/transformTest_cfg.py (hunk 1 above)
Comment (makortel) [reply FWCore/Integration/test/transformTest_cfg.py:32]: Right, basically from anywhere in the data dependence chain of the module registering the transformation call.

File: FWCore/Framework/src/ProductResolvers.cc (hunk 2 above)
Comment (makortel) [reply FWCore/Framework/src/ProductResolvers.cc:525]: The `putProduct()` call from `Event::commit_()`
https://github.com/cms-sw/cmssw/blob/ff2213a4904f42937039cad45098572009abdc40/FWCore/Framework/src/Event.cc#L170-L175
will always have a null wrapper. The Transformer itself calls `putProduct()` via `EventPrincipal::put()`
https://github.com/cms-sw/cmssw/blob/ff2213a4904f42937039cad45098572009abdc40/FWCore/Framework/src/EventPrincipal.cc#L209
https://github.com/cms-sw/cmssw/blob/ff2213a4904f42937039cad45098572009abdc40/FWCore/Framework/src/EventForTransformer.cc#L22-L26

Written that, I see now an alternative to customizing `putProduct()` that would be to customize (in some way) `productResolved()` to return `true` for Transformer. The `productResolved()` seems to be called only in `{Event,LuminosityBlock,Run,ProcessBlock}::commit_()`, but on a first thought "lying" in `productResolved()` sounds conceptually worse than leaving the status unaltered in `putProduct()`.

**Comment by makortel** [comment]
I'm coming to the conclusion that this is the least bad (quick) way, and given the "plan" to revisit the transformation ability in a context of more general memory space handling, and revisiting the Event system internals to make them more flexible, I would not spend much more time with this fix now.

**Comment by makortel** [comment]
@cmsbuild, please test

**Comment by makortel** [comment]
Another incarnation of #39803

**Comment by makortel** [comment]
+core

**Comment by perrotta** [comment]
+1

**Pull request merged**